### PR TITLE
corp: Add vectorized pow

### PR DESCRIPTION
**Pull request by gauss1314**
Signed-off-by: gauss <gauss1314@gmail.com>

PCP #5751 

###  What have you changed?

Add vectorized pow

###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?
- Unit test
- Integration test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

### CODE REVIEW

**Comment by gauss1314** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/53

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 1]
```diff
@@ -304,6 +305,22 @@ fn cot(arg: &Option<Real>) -> Result<Option<Real>> {
     }
 }
 
+#[inline]
+#[rpn_fn]
+fn pow(lhs: &Option<Real>, rhs: &Option<Real>) -> Result<Option<Real>> {
+    match (lhs, rhs) {
+        (Some(lhs), Some(rhs)) => {
+            let pow = (lhs.into_inner()).pow(rhs.into_inner());
+            if pow.is_infinite() || pow.is_nan() {
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: I don't think NaN results in overflow error.

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 2]
```diff
@@ -304,6 +305,22 @@ fn cot(arg: &Option<Real>) -> Result<Option<Real>> {
...
 }
 
+#[inline]
+#[rpn_fn]
+fn pow(lhs: &Option<Real>, rhs: &Option<Real>) -> Result<Option<Real>> {
+    match (lhs, rhs) {
+        (Some(lhs), Some(rhs)) => {
+            let pow = (lhs.into_inner()).pow(rhs.into_inner());
+            if pow.is_infinite() {
+                Err(Error::overflow("DOUBLE", format!("{}.pow({})", lhs, rhs)).into())
```
Comment (iosmanthus) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: I think we should handle overflow with an `EvalContext`

Comment (gauss1314) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: Shoud I add an input param like `&mut EvalContext `?

**Comment by gauss1314** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/53

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 3]
```diff
@@ -879,6 +896,39 @@ mod tests {
             .is_err());
     }
 
+    #[test]
+    fn test_pow() {
```
Comment (rleungx) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: How about adding some overflow test cases?

**Comment by gauss1314** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/53

File: components/tidb_query/src/rpn_expr/impl_math.rs (hunk 2 above)
Comment (gauss1314) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: Hi, I have added `EvalContext` handle overflow, please take a look. Thanks

**Comment by breeswish** [comment]
I don't think your implementation is correct according to the integration-copr-test result.

**Comment by gauss1314** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/53

**Comment by gauss1314** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/53

File: components/tidb_query/src/rpn_expr/impl_math.rs (hunk 2 above)
Comment (gauss1314) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: Hi, I notice there are many vectorized methods  handle overflow without `EvalContext`, so it seems we don't have a common way about error handing.
And when I add `EvalContext` in `pow` function, the push down test failed and I don't know how to fix it:

 NoPushDown Output: 
Error 1690: DOUBLE value is out of range in 'pow(1983, 12795)'

PushDownWithoutVec Output: 
Error 1690: DOUBLE value is out of range in '1983.pow(12795)'

PushDownWithVec Output: 
field1	field2
0	NULL
0	NULL
0	NULL
0	NULL
0	NULL
0	NULL
0	NULL
0	NULL
0	NULL
0	NULL
0	NULL

**Comment by gauss1314** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/53

**Comment by gauss1314** [comment]
...merge failed.
But I can't get any useful error message from the CI test log,  anybody knows what happened?


> [2019-11-23T12:58:53.598Z] thread 'cases::test_merge::test_node_merge_restart_after_apply_premerge_before_apply_compact_log' panicked at 'response header { error { message: "region 1000 not found" region_not_found { region_id: 1000 } } } has error', components/test_raftstore/src/cluster.rs:691:13
[2019-11-23T12:58:53.598Z] stack backtrace:

It seems have no contacts  with this PR....

**Comment by breeswish** [comment]
@gauss1314 Don't worry, it is an unstable test.

**Pull request merged**